### PR TITLE
Allow arbitrary expressions in attribute_path.

### PR DESCRIPTION
**Pull request by Fuuzetsu**
This allows more advanced attributes to be used. Motivating use-case
is ghcWithPackages.

### CODE REVIEW

File: nixpkgs/nixpkgs.bzl [hunk 1]
```diff
@@ -25,19 +25,27 @@ def _nixpkgs_package_impl(ctx):
...
+    path = "-I nixpkgs={0}".format(ctx.attr.path)
 
   attr_path = ctx.attr.attribute_path or ctx.name
-  res = ctx.execute(["nix-build", path, "-A", attr_path, "--no-out-link"])
+  # Use -E instead of -A: this way we can use build arbitrary
+  # attributes rather than being limited to what -A accepts. An
+  # example of this is haskellPackages.ghcWithPackages which is
+  # unusable with -A.
+  res = ctx.execute(["nix-build", path, "-E",
+                     "(import <nixpkgs> {{}}).{0}".format(attr_path),
```
Comment (mboes) [inline nixpkgs/nixpkgs.bzl]: Also, if an attribute path is now really an expression, then the name should be changed and it should be documented.

**Comment by Fuuzetsu** [comment]
```
[shana@lenalee:~/programming/rules_nixpkgs]$ nix-shell -p bazel
these paths will be fetched (136.46 MiB download, 243.04 MiB unpacked):
  /nix/store/0s51xhmdnpy3kzndind3a9dgbdskw4p1-openjdk-8u152b16
  /nix/store/nlinmdc3fzh5nnh41bj4ydswqndn1cv5-openjdk-8u152b16-jre
  /nix/store/wl9x15jfc56rn4l0xbbb90pdaw56w3ac-bazel-0.8.0
copying path '/nix/store/nlinmdc3fzh5nnh41bj4ydswqndn1cv5-openjdk-8u152b16-jre' from 'http://cache.nixos.org'...
copying path '/nix/store/0s51xhmdnpy3kzndind3a9dgbdskw4p1-openjdk-8u152b16' from 'http://cache.nixos.org'...
copying path '/nix/store/wl9x15jfc56rn4l0xbbb90pdaw56w3ac-bazel-0.8.0' from 'http://cache.nixos.org'...

[nix-shell:~/programming/rules_nixpkgs]$ bazel clean --expunge
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.

[nix-shell:~/programming/rules_nixpkgs]$ bazel run //tests:run-hello
.........
INFO: Analysed target //tests:run-hello (16 packages loaded).
INFO: Found 1 target...
Target //tests:run-hello up-to-date:
  bazel-bin/tests/run-hello
INFO: Elapsed time: 5.417s, Critical Path: 0.02s
INFO: Build completed successfully, 4 total actions

INFO: Running command line: bazel-bin/tests/run-hello external/hello/nix/bin/hello
Executing:  external/hello/nix/bin/hello
Hello, world!
```

???

**Comment by Fuuzetsu** [comment]
I changed the implementation a bit to allow arbitrary expressions and added a couple of tests. I hope they pass now.

**Comment by Fuuzetsu** [comment]
Hm, I think nix 1.12 is needed for `-I` to nix-build, not great…

**Comment by mboes** [comment]
> ???

The oracle here is CircleCI. Green on master but red on this branch. Can't merge until this one goes green.

**Comment by Fuuzetsu** [comment]
Turned out to be dumb mistake, should be green, PATAL.

File: nixpkgs/nixpkgs.bzl [hunk 2]
```diff
@@ -15,6 +15,16 @@ nixpkgs_git_repository = repository_rule(
   local = False,
 )
 
+def _mk_build_expression(ctx):
+  """Generate a nix expression that picks a package from nixpkgs.
+  """
+  if ctx.attr.attribute and ctx.attr.expression:
+    fail("'attribute' and 'expression' are mutually exclusive.")
```
Comment (mboes) [inline nixpkgs/nixpkgs.bzl]: I don't think the two should be mutually exclusive. `-A` and `-E` are not mutually exclusive flags to `nix-build`. When used together, `-E e -A some.path` means `e.some.path`. We should reproduce the same semantics here.

**Comment by Fuuzetsu** [comment]
Addressed. Didn't realise you could combine -E and -A.

**Comment by mboes** [comment]
Note: for now we're documenting attributes in README, not using inline docstrings.

**Pull request merged**